### PR TITLE
Gossipsub support

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Hi!

I am new to this project and to the internals of how ipfs works - but I want to use gossipsub and this crate does not yet have support for it, from what I see. So here's my take on implementing it.

Please note that this is a very early draft - so you can stop me from doing anything idiotic before I drown myself in this! :laughing: 

Please tell me if the direction I am heading towards is right - or tell me how to do better!

### Checklist (can be deleted from PR description once items are checked)

- [ ] **New** code is “linted” i.e. code formatting via rustfmt and language idioms via clippy
- [ ] There are no extraneous changes like formatting, line reordering, etc. Keep the patch sizes small!
- [ ] There are functional and/or unit tests written, and they are passing
- [ ] There is suitable documentation. In our case, this means:
    - [ ] Each command has a usage example and API specification
    - [ ] Rustdoc tests are passing on all code-level comments
    - [ ] Differences between Rust’s IPFS implementation and the Go or JS implementations are explained
- [ ] Additions to CHANGELOG.md files

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
No need to approve the workflow run! I just want you to either tell me that I'm writing bullshit here or that the direction is okay and what I should do to make this go into a mergeable direction! :laughing:

File: src/lib.rs [hunk 1]
```diff
@@ -273,6 +273,11 @@ enum IpfsEvent {
     PubsubPublish(String, Vec<u8>, OneshotSender<()>),
     PubsubPeers(Option<String>, OneshotSender<Vec<PeerId>>),
     PubsubSubscribed(OneshotSender<Vec<String>>),
+    GossipsubSubscribe(String, OneshotSender<Option<SubscriptionStream>>),
+    GossipsubUnsubscribe(String, OneshotSender<bool>),
+    GossipsubPublish(String, Vec<u8>, OneshotSender<()>),
+    GossipsubPeers(Option<String>, OneshotSender<Vec<PeerId>>),
+    GossipsubSubscribed(OneshotSender<Vec<String>>),
```
Comment (koivunej) [inline src/lib.rs:280]: Last I looked, gossipsub is a superset of the floodsub so I think the old entries in IpfsEvent and generally any api just should switch to using gossipsub as the backend. At least it should spare you from writing the boilerplate methods.

File: src/p2p/behaviour.rs [hunk 2]
```diff
@@ -532,6 +538,10 @@ impl<Types: IpfsTypes> Behaviour<Types> {
         &mut self.pubsub
     }
 
+    pub fn gossipsub(&mut self) -> &mut Gossipsub {
+        &mut self.gossipsub
+    }
+
```
Comment (koivunej) [inline src/p2p/behaviour.rs:544]: This would become unnecessary with the upgrade of pubsub to use gossipsub.

**Review by koivunej — CHANGES_REQUESTED** [review]
It's a nice start, but as mentioned in the comments, just upgrading the current src/p2p/pubsub.rs from floodsub to gossipsub would be the preferred way, hopefully leaving you to do only minimal boilerplate changes.

One issue I can't remember checking with the gossipsub implementation is that are the locally sent messages possible to be echoed in subscriptions. This behaviour at least was expected by the ipfs-interface test suite (the js tests) but it is very much a "cli only" functionality which might not require changes to gossipsub, but would have to be shimmied some other way. However, that's something to worry in the later phases or next PR.

**Comment by dariusc93** [comment]
Any update on this? If OP isnt interested I may try to take up on bringing gossip. Still reading over specs, libp2p code but it should be a simple implementation to replace floodsub but is there intentions to keep/use both protocols?

**Comment by koivunej** [comment]
> Still reading over specs, libp2p code but it should be a simple implementation to replace floodsub but is there intentions to keep/use both protocols?

@dariusc93 No intention to keep both as gossipsub impl in libp2p handles floodsub as well. A simple implementation up to the point of the local echo or that's what I've been assuming, but I haven't checked the gossipsub code in a while.

**Comment by matthiasbeyer** [comment]
I no longer have time to work on this, sorry. If someone wants to pick this up, feel free to. If not, this can of course be closed.

**Comment by dariusc93** [comment]
> I no longer have time to work on this, sorry. If someone wants to pick this up, feel free to. If not, this can of course be closed.

No need to be sorry. Already picked up on this in #507 though still investigating a couple of things (some related to gossipsub and others not so much such as mdns, bootstrap, etc).

**Pull request closed without merging**